opencog/atoms/core/Variables.cc

```diff
@@ -462,76 +465,33 @@ Handle FreeVariables::substitute_scoped(const Handle& term,
 	quotation.update(ty);
 
+	// If the term is a scope the index map might change, to avoid copy
+	// we either point to the original map, or the new one.
+	const IndexMap* index_map_ptr = &index_map;
+
 	if (unquoted and nameserver().isA(ty, SCOPE_LINK))
 	{
-		// Perform alpha-conversion duck-n-cover.  We don't actually need
-		// to alpha-convert anything, if we happen to encounter a bound
-		// variable that happens to have the same name as a free variable.
-		// Instead, the bound variable simply "hides" the free variable
-		// for as long as the bound variable is in scope. We hide it by
-		// removing it from the index.
-		ScopeLinkPtr sco(ScopeLinkCast(term));
-
-		const Variables& vees = sco->get_variables();
-		bool alpha_hide = false;
-		for (const Handle& v : vees.varseq)
-		{
-			IndexMap::const_iterator idx = index_map.find(v);
-			if (idx != index_map.end())
-			{
-				alpha_hide = true;
-				break;
-			}
-		}
+		// Perform alpha-conversion duck-n-cover.
 
-		// Hiding is expensive, so perform it only if we really have to.
-		if (alpha_hide)
-		{
-			// Make a copy... this is what's computationally expensive.
-			IndexMap hidden_map = index_map;
-
-			// Remove the alpha-hidden variables.
-			for (const Handle& v : vees.varseq)
-			{
-				IndexMap::const_iterator idx = hidden_map.find(v);
-				if (idx != hidden_map.end())
-				{
-					hidden_map.erase(idx);
-				}
-			}
-
-			// Also remove everything that is not a variable.
-			// The map will, in general, contain terms that
-			// contain alpha-hidden variables; those also have
-			// to go, or they will mess up the substitution.
-			for (auto it = hidden_map.begin(); it != hidden_map.end();)
-			{
-				Type tt = it->first->get_type();
-				if (tt != VARIABLE_NODE and tt != GLOB_NODE)
-				{
-					it = hidden_map.erase(it);
-				}
-				else
-				{
-					++it;
-				}
-			}
+		// If a substituting value is equal to a variable of that scope,
+		// then alpha-convert the scope to avoid variable name
+		// collision. Loop in the rare case the new names collide.
+		while (must_alpha_convert(term, args))
+			term = ScopeLinkCast(term)->alpha_convert();
 
+		// Hide any variables of the scope that are to be substituted,
+		// that is remove them from the index.
+		if (must_alpha_hide(term, index_map))
+		{
+			IndexMap hidden_map = alpha_hide(term, index_map);
 
 			// If the hidden map is empty, then there is no more
 			// substitution to be done.
 			if (hidden_map.empty())
 				return term;
 
-			// Recursively fill out the subtrees. Same as below, but
-			// using the alpha-renamed variable index map.
-			HandleSeq oset;
-			for (const Handle& h : term->getOutgoingSet())
-			{
-				oset.emplace_back(substitute_scoped(h, args, silent,
-				                                    hidden_map, quotation));
-			}
-			return createLink(std::move(oset), term->get_type());
+			// Otherwise the new map must be passed down below
+			index_map_ptr = &hidden_map;
```

Comment:
This risks a memory-access fault, or other corruption, because `hidden_map` goes out of scope, and it's destructor runs, with the closing brace below. I think it's some happy accident that it didn't die ... probably because its on the stack, and the dtor wont run until the stack is popped.

Reply:
I fixed this in f84c6d84cd619e8c11105bf2abd1d7500efff479 in master.